beancount-parser/src/error.rs

```diff
@@ -0,0 +1,186 @@
+            }
+        }
+        write!(f, " at line {} column {}", self.location.0, self.location.1)
+    }
+}
+
+impl Error for ParseError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        self.source.as_ref().map(|x| &**x)
+    }
+}
+
+impl ParseError {
+    pub(crate) fn invalid_state<T: ToString>(msg: T) -> ParseError {
+        ParseError {
+            kind: ParseErrorKind::InvalidParserState {
+                message: msg.to_string(),
+            },
+            location: (0, 0),
+            source: None,
+        }
+    }
+
+    pub(crate) fn invalid_state_with_span<T: ToString>(msg: T, span: Span) -> ParseError {
+        ParseError {
+            kind: ParseErrorKind::InvalidParserState {
+                message: msg.to_string(),
+            },
+            location: span.start_pos().line_col(),
+            source: None,
+        }
+    }
+
+    pub(crate) fn decimal_parse_error(err: rust_decimal::Error, span: Span) -> ParseError {
+        ParseError {
+            kind: ParseErrorKind::InvalidInput {
+                message: format!("failed to parse number: {}", err),
+            },
+            location: span.start_pos().line_col(),
+            source: None,
+        }
+    }
+}
+
+impl From<pest::error::Error<Rule>> for ParseError {
+    fn from(err: pest::error::Error<Rule>) -> Self {
+        let err = err.renamed_rules(|rule| {
+            match *rule {
+                Rule::EOI => "end of input",
+                Rule::WHITESPACE => "whitespace",
+                Rule::COMMENT => "comment",
+                Rule::bool => "boolean value",
+                Rule::indent => "indentation",
+                Rule::eol => "end of line",
+                Rule::key => "key",
+                Rule::value => "value",
+                Rule::key_value => "key-value pair",
+                Rule::key_value_line => "key-value line",
+                Rule::eol_kv_list => "newline followed by key-value line",
+                Rule::year => "4-digit year",
+                Rule::month => "2-digit month",
+                Rule::day => "2-digit day",
+                Rule::date_separator => "date separator ('-' or '/')",
+                Rule::date => "date",
+                Rule::num => "number",
+                Rule::int => "integer",
+                Rule::separated_int => "integer with separators",
+                Rule::operation => "numeric operator",
+                Rule::add => "'+'",
+                Rule::subtract => "'-'",
+                Rule::multiply => "'*'",
+                Rule::divide => "'/'",
+                Rule::num_expr => "numeric expression",
+                Rule::num_prefix => "numeric prefix",
+                Rule::term => "numeric expression term",
+                Rule::amount => "an amount",
+                Rule::double_quote => "double quotation mark",
+                Rule::quoted_str => "a quoted string",
```

Comment:
In the name of consistency, does this need to be prefixed with `a ...`?  Or could it just be `quoted_string`?  I think prefixing this way makes sense for certain rules (e.g. `Rule::quoted_char` below), but maybe less so for this one?

It seems throughout this `Impl` that some things are prefixed with a/an and others aren't.  It could be good to be consistent here either way, when possible.

Reply:
Yeah, I haven't been consistent in that regard. I think I'll remove the prefix in most places where it isn't strictly necessary.

For many of these rules it doesn't seem possible to get them in an actual error message because they are part of another atomic rule, e.g. `double_quote`.